hydromt/raster.py

```diff
@@ -150,63 +174,91 @@ def full_from_transform(
         attrs=attrs,
         crs=crs,
         lazy=lazy,
+        shape=shape,
+        dims=dims,
     )
     return da
 
 
 class XGeoBase(object):
     """This is the base class for the GIS extensions for xarray"""
 
-    def __init__(self, xarray_obj):
+    def __init__(self, xarray_obj: xr.DataArray | xr.Dataset) -> None:
         self._obj = xarray_obj
         # create new coordinate with attributes in which to save x_dim, y_dim and crs.
         # other spatial properties are always calculated on the fly to ensure consistency with data
         if GEO_MAP_COORD not in self._obj.coords:
             self._obj.coords[GEO_MAP_COORD] = xr.Variable((), 1)
 
     @property
-    def attrs(self):
+    def attrs(self) -> dict:
         """Return dictionary of spatial attributes"""
         return self._obj.coords[GEO_MAP_COORD].attrs
 
-    def set_attrs(self, **kwargs):
+    def set_attrs(self, **kwargs) -> None:
         """Update spatial attributes. Usage raster.set_attr(key=value)."""
         self._obj.coords[GEO_MAP_COORD].attrs.update(**kwargs)
 
-    def get_attrs(self, key, placeholder=None):
+    def get_attrs(self, key, placeholder=None) -> Any:
         """Return single spatial attribute."""
         return self._obj.coords[GEO_MAP_COORD].attrs.get(key, placeholder)
 
     @property
-    def crs(self):
+    def crs(self) -> CRS:
         """Return Coordinate Reference System as :py:meth:`pyproj.CRS` object."""
         return self._obj.rio.crs
 
     @property
-    def x_dim(self):
+    def x_dim(self) -> str:
         """Return the x dimension name"""
         if self.get_attrs("x_dim") not in self._obj.dims:
             self.set_spatial_dims()
         return self.attrs["x_dim"]
 
     @property
-    def y_dim(self):
+    def y_dim(self) -> str:
         """Return the y dimension name"""
         if self.get_attrs("y_dim") not in self._obj.dims:
             self.set_spatial_dims()
         return self.attrs["y_dim"]
 
     @property
-    def xcoords(self):
+    def xcoords(self) -> xr.IndexVariable:
         """Return the x coordinates"""
-        return self._obj[self.x_dim]
+        xcoords = self._obj[self.x_dim]
+        if self.x_dim not in self._obj.coords:
+            for key in list(self._obj.coords.keys()):
+                if key.startswith(self.x_dim):
+                    xcoords = self._obj.coords[key]
+                    break
+        if xcoords.ndim == 2 and list(xcoords.dims).index(self.x_dim) != 1:
```

Comment:
The idea is to make sure that we get good error messages if the order of dimensions is wrong (only for rotated) or the size of a dimension is smaller than 2 (you need at least two cells in each direction to make a grid - currently you get really weird errors). I've put it here because I use xcoords and ycoords to calculate the resolution, origin and rotation which are use for the transform. I'll double check if this is the right place / how it compares to `_check_dimensions`